### ISSUE TITLE
Lambda Variable Capture qn

### ISSUE THREAD

**Issue by Elyxpenguin** [nus-cs2030/2122-s1#262]
Hi, this is a qn from a pyp. Why is it that there is no error for r2? What is the difference with or without the static keyword for the variable y?

<img width="835" alt="Screenshot 2021-11-16 at 1 24 16 AM" src="https://user-images.githubusercontent.com/92641940/141826592-ee3cc105-45b5-400a-a1c8-d29cf805a6aa.png">

**Comment by mewrion** [comment]
> What is the difference with or without the static keyword for the variable y?

Static keyword means it is declared as the class instance attribute -> i.e. all objects of type Foo will have the attribute y, and the value will be 1

> Why is it that there is no error for r2?

In a lambda, it behaves similarly to a local inner class. If I'm not wrong, it can only capture variables that are effectively final.
(Someone correct me if I'm wrong, but I think this was implemented to prevent confusing situations when you modify the local variable in the method, which confuses the inner class dependent on that local variable)

**Comment by RussellDash332** [comment]
Adding to @mewrion, suppose y is now `final`, then we'll get an error.

**Comment by ff1Zzd** [comment]
> Adding to @mewrion, suppose y is now `final`, then we'll get an error.

Sorry I am a bit confused here, I tot y is supposed to be effectively final here? Or only variables initiated inside the inner class need to be effectively final?

**Comment by mewrion** [comment]
If you declare it as final, you are not able to modify it in the statement `y = y + 1`, considering you are actually indicating that this variable should be constant. 

In this context, you're right that it should be effectively final, but I think this attribute relates to the Memory Model. `static variables` are stored on the metaspace/non-heap, but in `bar()`, x variable is stored on the stack. 

(Again I might be wrong here, but from some quick googling, the issue is that lambda is not **capturing the local variable value**, but simply making a copy of it on the stack. The external modification to x is not visible to the lambda, so it throws the error)

**Comment by huzaifa1712** [comment]
I don't get why there's no error for r2 either. It seems as if the fact that y is immediately changed in the next statement would make it not effectively final, the same way changing x in the other function causes the error (if you remove the x = x + 1 statement it compiles without error).

**Comment by huzaifa1712** [comment]
After some googling, this stack overflow thread seems to give a different explanation for variable capture: https://stackoverflow.com/questions/32272713/lambda-expression-and-variable-capture

What I understand from reading this is that lambdas and other inner classes can capture the reference to the enclosing class, which allows us to do something like Foo.this.y to refer to static or other instance variables. So for this reason baz() has no issue with y being modified after, because it's just using the reference to the outer Foo class to refer to y.

But for bar(), x is a local variable that has not been declared final and neither is it effectively final since it's being modified after the lambda declaration. So it would cause confusion about which value to use.

Though I'm still kind of confused as to why it can't just capture and copy the local variable as it was at the time of declaration even if it's not final/effectively final. As in, I understand that Java throws an error for this but what's the reasoning..

**Comment by mewrion** [comment]
> Though I'm still kind of confused as to why it can't just capture and copy the local variable as it was at the time of declaration even if it's not final/effectively final. As in, I understand that Java throws an error for this but what's the reasoning..

What do you mean? In the thread you sent, it does explain why it can't copy the variable at the point of declaration. It becomes confusing, becomes after declaring the method, you modify the local variable again.

Consider this scenario

```
Runnable bar() {
    boolean test = true
    (
    Runnable r1 =() -> {
        while (test) {
        // do operation...
        }
    }
    test = false;
    return r1;
}
```
Here, there is some ambiguity as to whether your while loop actually sees the change to the boolean value of test. In Java, I'm guessing in order to circumvent such scenarios, it impose the final restriction

**Comment by huzaifa1712** [comment]
> Here, there is some ambiguity as to whether your while loop actually sees the change to the boolean value of test. In Java, I'm guessing in order to circumvent such scenarios, it impose the final restriction

Right, I see why there's ambiguity there. But it's just that, instead of just not allowing this situation to even compile, another way would be to make the lambda copy the variable being referenced at the time of declaration and essentially treat it as final even if it is not. So in this case test would be copied as true into the lambda and it would lead to an infinite loop.

If you try this sort of thing in Python it does yet another approach, which is essentially to track the changes in the variable the same way you can track changes in outer instance/static variables. So the loop in r1 wouldn't run at all because test would be false.

I just mean that I think there's more approaches than just not allowing it at all, and I was wondering if there's a good explanation for Java's approach over alternatives.

**Comment by DerenC** [comment]
![image](https://user-images.githubusercontent.com/65486901/142731007-0bb93259-acb2-441d-a0b4-062400fd58d7.png)
https://www.baeldung.com/java-lambda-effectively-final-local-variables
